### DISCUSSION TITLE
Iterating over More Complex Data

### DISCUSSION THREAD

**Discussion by ghost** [leptos-rs/leptos#2689] [Q&A] [answered]
Hello. The book [touches](https://book.leptos.dev/view/04b_iteration.html) on the question of how to write complex client applications in leptos. There are 3 options for dynamic collections with complex structures (which make up any even slightly complex client web application). Of which only option No. 2 (Nested Signals) seems good. I'm sorry, but I just don't understand how to write complex web applications in Leptos. I only had experience with React and Dioxus, and there dynamic collections are described directly and naturally.

1. I would like to know if there are examples of applications that have nested signals and interaction with the server. (I looked at the hackernews example, but it’s just a static rendering of data received from the server)

2. Is it normal to create trees of nested structures whose fields are signals? Will there be memory leaks or other problems?

3. Let's say I'm writing an application in the Optimistic UI style. An actions log comes from the server, which I sequentially apply to the data model on the client (nested structs with has signals). Is it possible to temporarily "disable" page rendering until I apply the entire stack of actions to the model, and then draw all the changes to the screen?

**Comment by gbj** [comment]
The `counters` and `todomvc` examples both include nested signals and describe their caveats, including the potential for memory leaks if you don't dispose of nested signals correctly. 

I've done some pretty significant work to minimize the need to manually dispose signals in 0.7, so compare to those examples on the `leptos_0.7` branch. 

Oh I've also done some very major work in 0.7 on nested reactivity in general. See the `stores` example for that. 

For 3) the answer is "yes" and the mechanics just depend on the details.

**Reply by ghost** [reply]
> For 3) the answer is "yes" and the mechanics just depend on the details

Thanks. Can you at least give me a hint on how to do this?)

**Reply by gbj** [reply, marked as answer]
Sure! So, all rendering is handled individually by effects that read from signals -- when you pass something reactive into the view (whether that's a single signal or `move || ...` closure that reads from a signal), the renderer wraps it in a render effect that will react to any signal reads inside it.

By default every signal read is tracked and every signal write notifies any subscribers, including the renderer, to ensure consistency between the data and the view.

If you explicitly want to "tear" (i.e., have the view show something that is different from what the data contains) you can do that by using untracked reads (`.get_untracked()` etc) combined with some "trigger" signal that you will notify when you want them to update.

```rust
let trigger = Trigger::new();
let some_signal = RwSignal::new(13);

view! { 
<p>
  {move || {
    trigger.track();
    some_signal.get()
   }}
  </p>
  <button on:click=move |_| some_signal.update(|n| *n += 1)>"+1"</button>
  <button on:click=move |_| trigger.notify()>"Show changes"</button>
```
(I just typed this in, apologies for any errors but hopefully it illustrates what I mean)

If this *isn't* what you mean, then give more details about what you're talking about and I'm happy to try to help.
